### ISSUE TITLE
New SelectField implementation

### ISSUE THREAD

**Issue by kvesteri** [wtforms/wtforms#44]
## Problem

Now that WTForms 2.0 is approaching I think its time to rethink the whole implementation of SelectFields. Also as wtforms.ext.sqlalchemy will be deprecated and the functionality will be moved to WTForms-Alchemy, I think its time to rethink QuerySelectFields.

The problems I see with the current implementation:
1. There can be only one value besides queried values in QuerySelectField (the blank value). One can not customize this value nor can he add more of these values in QuerySelectField. I've had cases where I needed multiple non-queried values for QuerySelectField. A classic example is a filter where one value represents 'No selection', one value represents 'All' and the other values are queried values.
2. SelectFields do not allow the rendering of optgroups
3. QuerySelectFields do not allow the rendering of optgroups
4. SelectFields do not allow lazy-evaluated choices (callables as choices)
## Proposal

SelectFields should use special Choices classes internally which encapsulate the logic behind these iterables. This would allow us to:
1. Get rid of QuerySelectField and QueryMultipleSelectField and instead use:
   
   ``` python
   SelectField(choices=QueryChoices(...))
   ```
2. The old constructor syntax would still work. For example:
   
   ``` python
   SelectField(choices=[(..), (..)]) 
   ```
3. The Choices classes would support nesting and the new SelectField would render nested choices as optgroups.
4. One could combine regular choices to QueryChoices.
   
   ``` python
   SelectField(
      choices=Choices(
          [
              ('', 'No selection'),
              ('all', 'All'),
          ]
      ) + QueryChoices(query_factor=lambda: session.query(User))
   )
   ```

A working example is available at: https://github.com/kvesteri/wtforms-components/blob/features/choices/wtforms_components/fields/select.py

An example implementation of QueryChoices:

``` python

def get_pk_from_identity(obj):
    cls, key = identity_key(instance=obj)
    return ':'.join(map(six.text_type, key))


def identity(func):
    return func


def labelize(func):
    if func is None:
        return identity
    elif isinstance(func, six.string_types):
        return operator.attrgetter(func)
    return func


class QueryChoices(Choices):
    def __init__(self, query_factory, get_pk=None, get_label=None):
        if get_pk is None:
            if not has_identity_key:
                raise Exception(
                    'The sqlalchemy identity_key function could not be '
                    'imported.'
                )
            self.get_pk = get_pk_from_identity
        else:
            self.get_pk = get_pk

        self.get_label = labelize(get_label)
        self.query_factory = query_factory
        self._object_list = None

    @property
    def object_list(self):
        if self._object_list is None:
            query = self.query_factory()
            self._object_list = list(
                (six.text_type(self.get_pk(obj)), obj) for obj in query
            )
        return self._object_list

    @property
    def choices(self):
        for pk, obj in self.object_list:
            yield pk, self.get_label(obj), obj
```

**Comment by azmeuk** [comment]
I wonder if this ticket is still relevant, though `optgroup` support is definitively a good idea.

**Comment by azmeuk** [comment]
`optgroup` support is addressed in #656
